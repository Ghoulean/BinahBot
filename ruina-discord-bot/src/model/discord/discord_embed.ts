/**
 * Object holding embed data.
 * 
 * See also: https://discord.com/developers/docs/resources/channel#embed-object
 */
export type DiscordEmbed = {
    title?: string;
    description?: string;
    color?: number;
    image?: {
        url: string;
    },
    footer?: {
        text: string;
        icon_url?: string;
    },
    author?: {
        name: string;
        url?: string;
        icon_url?: string;
    },
    fields: DiscordEmbedFields[];
}

export type DiscordEmbedFields = {
    name: string;
    value: string;
    inline?: boolean;
}
