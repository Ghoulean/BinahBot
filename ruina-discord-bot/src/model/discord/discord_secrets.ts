/**
 * Object holding Discord secrets stored in AWS Secrets Manager
 * 
 */
export type DiscordSecrets = {
    applicationId: string;
    authToken: string;
    publicKey: string;
}
