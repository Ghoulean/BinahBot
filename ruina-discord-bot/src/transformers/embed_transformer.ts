import {
    DecoratedAbnoPage,
    DecoratedCombatPage,
    DecoratedKeyPage,
    DecoratedPassive,
    Die,
    Localization,
    LookupResult,
    Rarity,
    dieTypeToShortString,
} from "@ghoulean/ruina-common";
import { DisambiguationResults } from "../model/disambiguation_result";
import {
    DiscordEmbed,
    DiscordEmbedFields,
} from "../model/discord/discord_embed";
import { DiscordEmbedColors } from "../util/constants";

export class EmbedTransformer {
    private readonly s3BucketName: string;

    constructor(s3BucketName: string) {
        this.s3BucketName = s3BucketName;
    }

    // TODO: localize via _requestLocale
    public transformAbnoPage(
        abno: DecoratedAbnoPage,
        _requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number =
            abno.emotionSign > 0
                ? DiscordEmbedColors.AWAKENING_ABNO_PAGE
                : DiscordEmbedColors.BREAKDOWN_ABNO_PAGE;
        const abnoType: string =
            abno.emotionSign > 0 ? "Awakening" : "Breakdown";
        return {
            title: abno.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${abno.imagePath}`,
            },
            fields: [
                {
                    name: "Flavor text",
                    value: abno.flavorText,
                    inline: true,
                },
                {
                    name: "Effect",
                    value: abno.description,
                    inline: true,
                },
                {
                    name: "Bias",
                    value: String(abno.emotionRate),
                    inline: true,
                },
                {
                    name: "Type",
                    value: abnoType,
                    inline: true,
                },
                {
                    name: "Emotion level",
                    value: String(abno.emoLevel),
                    inline: true,
                },
                {
                    name: "Floor",
                    value: abno.floor,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformCombatPage(
        combat: DecoratedCombatPage,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(combat.rarity);
        const fields: DiscordEmbedFields[] = [
            {
                name: "Cost",
                value: String(combat.cost),
                inline: true,
            },
            {
                name: "Range",
                value: combat.range,
                inline: true,
            },
            {
                name: "Rarity",
                value: combat.rarity,
                inline: true,
            },
        ];
        if (combat.description) {
            fields.push({
                name: "Page Description",
                value: combat.description,
                inline: true,
            });
        }

        fields.push({
            name: "Dice",
            value: this.mapDie(
                combat.dice,
                combat.diceDescriptions,
                requestLocale
            ),
            inline: false,
        });
        return {
            title: combat.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${combat.imagePath}`,
            },
            fields: fields,
        };
    }
    // TODO: localize via _requestLocale
    public transformKeyPage(
        keyPage: DecoratedKeyPage,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(keyPage.rarity);

        return {
            title: keyPage.name,
            color: embedColor,
            fields: [
                {
                    name: "Rarity",
                    value: keyPage.rarity,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformPassive(
        passive: DecoratedPassive,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(passive.rarity);

        return {
            title: passive.name,
            color: embedColor,
            fields: [
                {
                    name: "Cost",
                    value: String(passive.cost),
                    inline: true,
                },
                {
                    name: "Rarity",
                    value: passive.rarity,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformDisambiguationPage(
        disambiguation: DisambiguationResults,
        _requestLocale: Localization
    ): DiscordEmbed {
        return {
            fields: [
                {
                    name: `**${disambiguation.query}** may refer to:`,
                    value: this.mapToIndentedList(
                        disambiguation.lookupResults.map(
                            (lookupResult: LookupResult) => {
                                return lookupResult.query;
                            }
                        )
                    ),
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public noResultsFoundEmbed(
        query: string,
        _requestLocale: Localization
    ): DiscordEmbed {
        return {
            description: `*No results found for \`${query}\`.*`,
            fields: [],
        };
    }

    private rarityToColor(rarity: Rarity): number {
        switch (rarity) {
            case Rarity.PAPERBACK:
                return DiscordEmbedColors.PAPERBACK_RARITY;
            case Rarity.HARDCOVER:
                return DiscordEmbedColors.HARDCOVER_RARITY;
            case Rarity.LIMITED:
                return DiscordEmbedColors.LIMITED_RARITY;
            case Rarity.OBJET_D_ART:
                return DiscordEmbedColors.OBJET_D_ART_RARITY;
        }
    }

    private mapDie(
        dice: Die[],
        diceDescriptions: string[],
        locale: Localization
    ): string {
        return this.mapToIndentedList(
            dice.map((die: Die, index: number) => {
                return `${dieTypeToShortString(die.type, locale)} ${die.min}-${
                    die.max
                } ${diceDescriptions[index]}`;
            })
        );
    }

    private mapToIndentedList(arr: string[]): string {
        return arr
            .map((str: string) => {
                return ` > - ${str}`;
            })
            .join("\n");
    }
}
