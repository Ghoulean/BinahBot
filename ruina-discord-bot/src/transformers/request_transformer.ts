import { Chapter } from "@ghoulean/ruina-common";
import * as __SPOILER_CONFIG from "../config/spoiler_config.json";
import {
    DiscordInteraction,
    DiscordInteractionOptions,
} from "../model/discord/discord_interaction";
import { Request, RequestCommandArgs } from "../model/request";
import { DiscordInteractionTypes } from "../util/constants";
import { Util } from "../util/util";

// TODO: consider spoiler config accessor?
const SPOILER_CONFIG = __SPOILER_CONFIG as { [key: string]: Chapter };

const DEFAULT_DISCORD_LOCALE = "en-US";

export class RequestTransformer {
    constructor() {}

    public transformInteractionToRequest(
        interaction: DiscordInteraction
    ): Request {
        const locale: string =
            interaction.locale ??
            interaction.guild_locale ??
            DEFAULT_DISCORD_LOCALE;
        if (!interaction.data) {
            throw new Error("Interaction data not found");
        }
        const commandArgs: RequestCommandArgs = interaction.data.options
            ? this.transformArgs(interaction.data.options)
            : {};
        return {
            command: interaction.data.name,
            commandArgs: commandArgs,
            interactionToken: interaction.token,
            locale: Util.deserializeDiscordLocale(locale),
            chapter: this.getChapter(interaction.channel_id)
        };
    }

    private transformArgs(
        discordInteractionOptions: DiscordInteractionOptions[]
    ): RequestCommandArgs {
        const retVal: RequestCommandArgs = {};
        for (const option of discordInteractionOptions) {
            retVal[option.name] = option.value;
        }
        return retVal;
    }

    private getChapter(channelId?: string): Chapter {
        return (
            (channelId && SPOILER_CONFIG[channelId]) ||
            Chapter.IMPURITAS_CIVITATIS
        );
    }
}
