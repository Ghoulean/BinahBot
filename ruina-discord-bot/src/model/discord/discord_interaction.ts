import { DiscordEmbed } from "./discord_embed";

/**
 * Object holding interaction data. This only reflects the information that the bot cares about; other fields are discarded.
 *
 * See also: https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object
 */
export type DiscordInteraction = {
    id: string;
    type: number;
    token: string;
    application_id: string;
    locale?: string;
    guild_locale?: string;
    channel_id?: string;
    data?: {
        id: string;
        name: string;
        options?: DiscordInteractionOptions[];
    };
    metadata: DiscordInteractionMetadata
};

export type DiscordInteractionOptions = {
    name: string;
    value: string;
};

export type DiscordInteractionMetadata = {
    timestamp: string;
    signature: string;
    jsonBody: string;
}

export type DiscordInteractionResponse = {
    type: number;
    data?: DiscordInteractionResponseMessage | DiscordInteractionResponseAutocomplete;
};

export type DiscordInteractionResponseMessage = {
    embeds?: DiscordEmbed[];
};

export type DiscordInteractionResponseAutocomplete = {
    choices?: DiscordInteractionOptions[];
};
