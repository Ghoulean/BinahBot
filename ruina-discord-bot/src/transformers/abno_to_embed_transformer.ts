import { DecoratedAbnoPage, Localization } from "@ghoulean/ruina-common";
import { DiscordEmbedColors } from "../util/constants";
import { DiscordEmbed } from "../model/discord/discord_embed";

export class AbnoToEmbedTransformer {
    private readonly s3BucketName: string;

    constructor(s3BucketName: string) {
        this.s3BucketName = s3BucketName;
    }

    // TODO: strongly type embeds
    // TODO: localize via abno.locale
    public transform(abno: DecoratedAbnoPage): DiscordEmbed {
        const embedColor: number = abno.emotionSign > 0 ? DiscordEmbedColors.AWAKENING_ABNO_PAGE : DiscordEmbedColors.BREAKDOWN_ABNO_PAGE;
        const abnoType: string = abno.emotionSign > 0 ? "Awakening" : "Breakdown";
        return {
            title: abno.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${abno.imagePath}`,
            },
            fields: [
                {
                    name: "Flavor text",
                    value: abno.flavorText,
                    inline: true,
                },
                {
                    name: "Effect",
                    value: abno.description,
                    inline: true,
                },
                {
                    name: "Bias",
                    value: String(abno.emotionRate),
                    inline: true,
                },
                {
                    name: "Type",
                    value: abnoType,
                    inline: true,
                },
                {
                    name: "Emotion level",
                    value: String(abno.emoLevel),
                    inline: true,
                },
                {
                    name: "Floor",
                    value: abno.floor,
                    inline: true,
                },
            ],
        };
    }
}
