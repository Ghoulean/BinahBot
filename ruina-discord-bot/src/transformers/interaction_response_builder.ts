import { DiscordEmbed } from "../model/discord/discord_embed";
import {
    DiscordInteractionResponse,
    DiscordInteractionResponseAutocomplete,
    DiscordInteractionResponseMessage
} from "../model/discord/discord_interaction";
import { DiscordInteractionTypes } from "../util/constants";

const PING_DISCORD_INTERACTION_TYPE = 1;
const MESSAGE_DISCORD_INTERACTION_TYPE = 4;
const AUTOCOMPLETE_RESULT_DISCORD_INTERACTION_TYPE = 8;

const PING_RESPONSE: DiscordInteractionResponse = {
    type: PING_DISCORD_INTERACTION_TYPE,
};

export class InteractionResponseBuilder {
    public buildResponse(
        requestType: number,
        data: unknown
    ): DiscordInteractionResponse {
        switch (requestType) {
            case DiscordInteractionTypes.PING:
                return PING_RESPONSE;
            case DiscordInteractionTypes.APPLICATION_COMMAND:
                return this.buildCommandResponse(data as DiscordEmbed);
            case DiscordInteractionTypes.APPLICATION_COMMAND_AUTOCOMPLETE:
                return this.buildAutocompleteResponse(data as string[]);
            default:
                throw new Error(
                    `Unrecognized or unsupported request type ${requestType}`
                );
        }
    }

    private buildCommandResponse(
        embed: DiscordEmbed
    ): DiscordInteractionResponse {
        const discordPayload: DiscordInteractionResponseMessage = {
            embeds: [embed],
        };
        return {
            type: MESSAGE_DISCORD_INTERACTION_TYPE,
            data: discordPayload,
        };
    }

    private buildAutocompleteResponse(
        choices: string[]
    ): DiscordInteractionResponse {
        const discordPayload: DiscordInteractionResponseAutocomplete = {
            choices: choices.map((s: string) => {
                return {
                    name: s,
                    value: s,
                };
            }),
        };
        return {
            type: AUTOCOMPLETE_RESULT_DISCORD_INTERACTION_TYPE,
            data: discordPayload,
        };
    }
}
