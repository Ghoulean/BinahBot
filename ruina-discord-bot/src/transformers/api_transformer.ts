import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { DiscordInteraction } from "../model/discord/discord_interaction";

const TIMESTAMP_HEADER = "x-signature-timestamp";
const SIGNATURE_HEADER = "x-signature-ed25519";

export class ApiTransformer {
    constructor() {}

    public transformApiRequestToInteractionPayload(
        event: APIGatewayProxyEvent
    ): DiscordInteraction {
        return {
            ...JSON.parse(event.body!) as DiscordInteraction,
            metadata: {
                timestamp: event.headers[TIMESTAMP_HEADER]!,
                signature: event.headers[SIGNATURE_HEADER]!,
                jsonBody: event.body!,
            },
        };
    }

    public transformDataBlobToApiResponse(blob: any): APIGatewayProxyResult {
        return {
            statusCode: 200,
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(blob),
        };
    }
}
