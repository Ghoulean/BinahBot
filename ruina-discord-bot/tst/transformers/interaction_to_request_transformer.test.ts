import { DiscordInteraction } from "../../src/model/discord/discord_interaction";
import { InteractionToRequestTransformer } from "../../src/transformers/interaction_to_request_transformer";

// TODO: reconsider testing strategy for spoiler channels
const DEBUG_SPOILER_CHANNEL = "270097753177194498";

const INTERACTION: DiscordInteraction = {
    id: "",
    token: "",
    application_id: "",
    data: {
        id: "",
        name: "",
    },
};

let interactionToRequestTransformer: InteractionToRequestTransformer;

beforeEach(() => {
    interactionToRequestTransformer = new InteractionToRequestTransformer();
});

test("should transform interaction to request", () => {
    expect(
        interactionToRequestTransformer.transform(INTERACTION)
    ).toMatchSnapshot();
});

test("should transform interaction options to command args", () => {
    const optionsInteraction: DiscordInteraction = {
        ...INTERACTION,
        data: {
            ...INTERACTION.data!,
            options: [
                {
                    name: "name",
                    value: "value",
                },
            ],
        },
    };
    expect(
        interactionToRequestTransformer.transform(optionsInteraction)
    ).toMatchSnapshot();
});

test("should prioritize locale correctly", () => {
    const guildLocaleInteraction: DiscordInteraction = {
        ...INTERACTION,
        guild_locale: "ko",
    };
    const localeInteraction: DiscordInteraction = {
        ...guildLocaleInteraction,
        locale: "ja",
    };
    expect(
        interactionToRequestTransformer.transform(guildLocaleInteraction)
    ).toMatchSnapshot();
    expect(
        interactionToRequestTransformer.transform(localeInteraction)
    ).toMatchSnapshot();
});

test("should set chapter level from spoiler config", () => {
    const interaction: DiscordInteraction = {
        ...INTERACTION,
        channel_id: DEBUG_SPOILER_CHANNEL
    };
    expect(
        interactionToRequestTransformer.transform(interaction)
    ).toMatchSnapshot();
});
