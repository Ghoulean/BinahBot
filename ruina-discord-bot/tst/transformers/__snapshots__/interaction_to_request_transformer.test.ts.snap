// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should prioritize locale correctly 1`] = `
{
  "chapter": 7,
  "command": "",
  "commandArgs": {},
  "interactionToken": "",
  "locale": "kr",
}
`;

exports[`should prioritize locale correctly 2`] = `
{
  "chapter": 7,
  "command": "",
  "commandArgs": {},
  "interactionToken": "",
  "locale": "jp",
}
`;

exports[`should set chapter level from spoiler config 1`] = `
{
  "chapter": 7,
  "command": "",
  "commandArgs": {},
  "interactionToken": "",
  "locale": "en",
}
`;

exports[`should transform interaction options to command args 1`] = `
{
  "chapter": 7,
  "command": "",
  "commandArgs": {
    "name": "value",
  },
  "interactionToken": "",
  "locale": "en",
}
`;

exports[`should transform interaction to request 1`] = `
{
  "chapter": 7,
  "command": "",
  "commandArgs": {},
  "interactionToken": "",
  "locale": "en",
}
`;
